File: 7_claudecode/expense-tracker/backend/analytics.py
"""Aggregation helpers for the summary endpoint and CSV export."""
from __future__ import annotations

import csv
import io
from collections import defaultdict
from datetime import datetime, timezone
from typing import List

from models import (
    Category,
    CategoryTotal,
    Expense,
    MonthlyTotal,
    Summary,
)


def compute_summary(expenses: List[Expense]) -> Summary:
    now = datetime.now(timezone.utc)
    month_key = f"{now.year}-{now.month:02d}"

    total = 0.0
    month_total = 0.0
    by_category: dict[Category, float] = defaultdict(float)
    by_month: dict[str, float] = defaultdict(float)

    for e in expenses:
        total += e.amount
        by_category[e.category] += e.amount
        m = e.date.strftime("%Y-%m")
        by_month[m] += e.amount
        if m == month_key:
            month_total += e.amount

    category_totals = sorted(
        (
            CategoryTotal(category=c, total=round(t, 2))
            for c, t in by_category.items()
            if t > 0
        ),
        key=lambda c: c.total,
        reverse=True,
    )

    monthly_totals = [
        MonthlyTotal(month=m, total=round(by_month[m], 2))
        for m in sorted(by_month)
    ][-6:]

    return Summary(
        total=round(total, 2),
        month_total=round(month_total, 2),
        count=len(expenses),
        average_per_expense=round(total / len(expenses), 2) if expenses else 0.0,
        top_category=category_totals[0] if category_totals else None,
        by_category=category_totals,
        by_month=monthly_totals,
    )


def expenses_to_csv(expenses: List[Expense]) -> str:
    buffer = io.StringIO()
    writer = csv.writer(buffer)
    writer.writerow(["Date", "Category", "Description", "Amount"])
    for e in sorted(expenses, key=lambda x: x.date, reverse=True):
        writer.writerow(
            [e.date.isoformat(), e.category.value, e.description, f"{e.amount:.2f}"]
        )
    return buffer.getvalue()

File: 7_claudecode/expense-tracker/frontend/lib/analytics.ts
import type { Category, Expense, ExpenseFilters } from "./types";
import { CATEGORIES } from "./types";

export interface CategoryTotal {
  category: Category;
  total: number;
}

export interface MonthlyTotal {
  /** YYYY-MM */
  month: string;
  /** Human label, e.g. "Mar 2026" */
  label: string;
  total: number;
}

export interface Summary {
  total: number;
  monthTotal: number;
  count: number;
  averagePerExpense: number;
  topCategory: CategoryTotal | null;
  byCategory: CategoryTotal[];
  byMonth: MonthlyTotal[];
}

function currentMonthKey(): string {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}`;
}

export function applyFilters(
  expenses: Expense[],
  filters: ExpenseFilters
): Expense[] {
  const term = filters.search.trim().toLowerCase();
  return expenses.filter((e) => {
    if (filters.category !== "All" && e.category !== filters.category) {
      return false;
    }
    if (filters.startDate && e.date < filters.startDate) return false;
    if (filters.endDate && e.date > filters.endDate) return false;
    if (term) {
      const haystack = `${e.description} ${e.category}`.toLowerCase();
      if (!haystack.includes(term)) return false;
    }
    return true;
  });
}

export function sortByDateDesc(expenses: Expense[]): Expense[] {
  return [...expenses].sort((a, b) => {
    if (a.date !== b.date) return a.date < b.date ? 1 : -1;
    return a.createdAt < b.createdAt ? 1 : -1;
  });
}

export function computeSummary(expenses: Expense[]): Summary {
  const monthKey = currentMonthKey();

  let total = 0;
  let monthTotal = 0;
  const categoryMap = new Map<Category, number>();
  const monthMap = new Map<string, number>();

  for (const e of expenses) {
    total += e.amount;
    categoryMap.set(e.category, (categoryMap.get(e.category) ?? 0) + e.amount);

    const m = e.date.slice(0, 7); // YYYY-MM
    monthMap.set(m, (monthMap.get(m) ?? 0) + e.amount);
    if (m === monthKey) monthTotal += e.amount;
  }

  const byCategory: CategoryTotal[] = CATEGORIES.map((category) => ({
    category,
    total: categoryMap.get(category) ?? 0,
  }))
    .filter((c) => c.total > 0)
    .sort((a, b) => b.total - a.total);

  const byMonth: MonthlyTotal[] = [...monthMap.entries()]
    .sort((a, b) => (a[0] < b[0] ? -1 : 1))
    .slice(-6)
    .map(([month, t]) => {
      const [y, mo] = month.split("-");
      const label = new Date(Number(y), Number(mo) - 1, 1).toLocaleDateString(
        "en-US",
        { month: "short", year: "numeric" }
      );
      return { month, label, total: t };
    });

  return {
    total,
    monthTotal,
    count: expenses.length,
    averagePerExpense: expenses.length ? total / expenses.length : 0,
    topCategory: byCategory[0] ?? null,
    byCategory,
    byMonth,
  };
}

export function expensesToCsv(expenses: Expense[]): string {
  const header = ["Date", "Category", "Description", "Amount"];
  const escape = (v: string) => `"${v.replace(/"/g, '""')}"`;
  const rows = expenses.map((e) =>
    [e.date, e.category, escape(e.description), e.amount.toFixed(2)].join(",")
  );
  return [header.join(","), ...rows].join("\n");
}
